source/dub/dependency.d

```diff
@@ -462,13 +462,36 @@ struct Dependency {
 		);
 	}
 
+	bool matches(ref const(Dependency) o) const @safe {
+		alias Merger = match!(
+			(const NativePath a, const NativePath b) => true, // treat all paths as equivalent (relative ones may likely diverge across projects in different directory depths)
+			(const NativePath a,       any         ) => true,
+			(      any         , const NativePath b) => false,
+
+			(const Repository a, const Repository b) => a == b,
+			(const Repository a,       any         ) => true,
+			(      any         , const Repository b) => false,
+
+			(const VersionRange a, const VersionRange b) {
+				if (a.matchesAny()) return b.isValid();
+				if (b.matchesAny()) return a.isValid();
+
+				VersionRange copy = a;
+				copy.merge(b);
+				return copy.isValid();
```

Comment:
Checking whether an actual version *range* matches another range is probably something else (I assume `VersionRange.merge()` returns the intersection). So this might need special-casing or asserting that `a` isn't a *range*, but a single version.